### PR TITLE
Feature/#204 consolidate search and multiselect

### PR DESCRIPTION
**Pull request by iguannalin**
See issue: [#204](https://github.com/mitre-attack/attack-navigator/issues/204)

### CODE REVIEW

File: nav-app/src/app/sidebar/sidebar.component.ts
Comment (isaisabel) [inline nav-app/src/app/sidebar/sidebar.component.ts:8]: This technically violates the custom component prefix recommendation for Angular, but so do _most if not all_ of the Navigator custom components so it's probably not worth fixing in just this one case. We may want to do a short code cleanup sprint later to fix this.

https://angular.io/guide/styleguide#component-custom-prefix

File: nav-app/src/app/techniques-search/techniques-search.component.ts
Comment (isaisabel) [inline nav-app/src/app/techniques-search/techniques-search.component.ts:6]: We may want to rename this component from `techniques-search` to something more general since we can search beyond just techniques in the interface now.

File: nav-app/src/app/techniques-search/techniques-search.component.scss
Comment (isaisabel) [inline nav-app/src/app/techniques-search/techniques-search.component.scss:72]: I believe the correct way to approach this is to disable ViewEncapsulation, ::ng-deep is deprecated:https://angular.io/guide/component-styles#deprecated-deep--and-ng-deep

File: nav-app/src/app/techniques-search/techniques-search.component.scss
Comment (isaisabel) [inline nav-app/src/app/techniques-search/techniques-search.component.scss:11]: ::ng-deep is deprecated (see other comment about this)

File: nav-app/src/app/techniques-search/techniques-search.component.ts
Comment (isaisabel) [inline nav-app/src/app/techniques-search/techniques-search.component.ts:70]: Args should probably be documented in doc comment, it's not inherently apparent what sortAllTechniques does without a doc comment.

File: nav-app/src/app/datatable/data-table.component.html
Comment (isaisabel) [inline nav-app/src/app/datatable/data-table.component.html:43]: The dropdown class should be removed here (which _should_ remove the arrow after the button) since it's not really a dropdown anymore.

File: nav-app/src/app/techniques-search/techniques-search.component.html
Comment (isaisabel) [inline nav-app/src/app/techniques-search/techniques-search.component.html:49]: Glad you kept the parent name as a prefix to the sub-technique name. I don't think I've mentioned it previously, but that's a standard we keep across the infrastructure -- if not shown hierarchically (not in the matrix or indented list, etc), the sub-technique name is always prepended with the parent name.

File: nav-app/src/app/techniques-search/techniques-search.component.ts
Comment (isaisabel) [inline nav-app/src/app/techniques-search/techniques-search.component.ts:71]: It seems like deprecated items are not being filtered out of the sidebar results:
image
"bash_profile and .bashrc Mitigation" is a deprecated mitigation and shouldn't be shown.

File: nav-app/src/app/techniques-search/techniques-search.component.html
Comment (isaisabel) [inline nav-app/src/app/techniques-search/techniques-search.component.html:3]: Label here should not be "search techniques" since it's searching all object types, not just techniques. Probably just "search" would be appropriate?

File: nav-app/src/app/techniques-search/techniques-search.component.ts
Comment (isaisabel) [reply nav-app/src/app/techniques-search/techniques-search.component.ts:71]: A quick way to tell (for mitigations specifically) is that if it ends with the word "mitigation" and the prefix looks like a technique name it's almost definitely a deprecated mitigation. There was an update which changed how mitigations behave and they used to follow that naming convention (`"[technique] Mitigation"`)

File: nav-app/src/app/techniques-search/techniques-search.component.ts
Comment (isaisabel) [reply nav-app/src/app/techniques-search/techniques-search.component.ts:70]: Should `items` be `BaseStix[]` instead of `any[]`? It seems like it will always be a type which inherits from [BaseStix](https://github.com/mitre-attack/attack-navigator/blob/master/nav-app/src/app/data.service.ts#L308).

File: nav-app/src/app/techniques-search/techniques-search.component.ts
Comment (isaisabel) [inline nav-app/src/app/techniques-search/techniques-search.component.ts:144]: I've not seen optimization like this before. Definitely something I will consider reusing myself in the future!

File: nav-app/src/app/techniques-search/techniques-search.component.ts
Comment (isaisabel) [inline nav-app/src/app/techniques-search/techniques-search.component.ts:171]: For the purposes of correctness `isStix` should be changed. Techniques are themselves STIX like all other objects. `isNonTechnique` (or `isTechnique` while inverting the if statement) would be better.

File: nav-app/src/app/techniques-search/techniques-search.component.ts
Comment (isaisabel) [inline nav-app/src/app/techniques-search/techniques-search.component.ts:214]: Again I think this is a misuse of the word Stix. Also STIX is misspelled in `deselectAllSitx`

**Review by isaisabel — CHANGES_REQUESTED** [review]
This is an _extremely_ snazzy UI. I am thoroughly impressed! The elevation of the blocks on the gray background is an excellent way to section the items and draw the eye to the important parts while keeping the sidebar visually separated from the matrix.

# Behavior
Behaviorally I have not found any notable bugs with what is selected. All of the edge cases I tried (involving the selection controls) checked out just fine. I have no concerns about the overall behavior of the feature.

We probably want to include an explicit close button on the sidebar since users might not know they can click the search button to close the sidebar. Maybe making it to the right of the search input (top right corner), inline with the input?

# Performance
I am a little bit concerned with the performance of the new highlighting code. Some quick performance analysis on the previous code shows it takes about 49ms to execute the mouseenter event and callbacks to highlight a cell. The new code takes anywhere from 70 ms to 100ms, which is _visibly_ less responsive. 

## Previous mouseenter performance:
Screen Shot 2021-06-09 at 3 45 02 PM

## Current mouseenter performance:
Screen Shot 2021-06-09 at 3 45 53 PM

We may want to look into possible performance enhancements to avoid visual latency in that part of the application. 

I should note that performance is always slightly better in the actual live deployment (presumably because of optimizations in the production build), so we may want to do more testing of the performance just in case I'm wrong in my method of measurement. Probably the best way to test performance is to check out the previous version (4.3) and run it in parallel on a different port and take the average of 10 observances of mousenter or something to that effect. This will produce closer conditions than in my testing.

File: nav-app/src/app/sidebar/sidebar.component.ts
Comment (iguannalin) [reply nav-app/src/app/sidebar/sidebar.component.ts:8]: Yeah, that's a good point. @clemiller and I talked about this today, and I think we should wait until both our issues have been merged into the develop branch before renaming the sidebar component.

**Comment by iguannalin** [comment]
> We may want to look into possible performance enhancements to avoid visual latency in that part of the application.
> 
> I should note that performance is always slightly better in the actual live deployment (presumably because of optimizations in the production build), so we may want to do more testing of the performance just in case I'm wrong in my method of measurement. Probably the best way to test performance is to check out the previous version (4.3) and run it in parallel on a different port and take the average of 10 observances of mousenter or something to that effect. This will produce closer conditions than in my testing.

@isaisabel, I looked into this and tested the performance of the _mouseenter_ event in 3 environments:
1. locally, develop branch (65.78ms)
2. locally, master branch (78.78ms)
2. locally, feature/#204 branch (76.59ms)
3. deployed[ v4.3](https://mitre-attack.github.io/attack-navigator/) (20.08ms)

I tested this by running the performance profiling, reloading the page, and hovering over the 4 techniques in the top left corner after the matrix loads. It seems to take consistently ~60-80ms for the mouseenter event when running locally (I ran these about 2-3 times), and significantly less at 20ms in the deployed version on the mitre-attack site. 

If I'm doing this right and understanding you correctly, I'm not sure if there's a significant lag in the mouseenter/hover because of the changes in this feature branch. One thing I do think concerns me about performance when working on this feature though, was the loading of all the techniques in the search component sidebar (since it now loads & searches techniques across domains and all techniques).

Screenshots:
1. Screen Shot 2021-06-16 at 9 32 16 AM
2. Screen Shot 2021-06-16 at 9 32 33 AM
3. Screen Shot 2021-06-16 at 9 39 24 AM
4. Screen Shot 2021-06-16 at 9 31 10 AM

**Review by isaisabel — CHANGES_REQUESTED** [review]
I actually have found one major bug in my second review. When switching between layers of different domains, the contents of the sidebar does not update properly until a query is added.

**Steps to reproduce:**
1. Open two layers, one Enterprise and one Mobile
2. Open the sidebar on both tabs
3. Switch back and forth between the two tabs. The counts and listed objects should change when switching tabs since the data under each domain is different, but this does not happen. Only when a query is inputted do the objects listed refresh.

Aside from that everything looks great, and the added denounce definitely makes the search input more performant.

**Review by isaisabel — APPROVED** [review]
Looks great! I pushed a commit which updates the help page documentation which was pretty old for this feature (it didn't mention mitigations which were not originally part of the interface and generally needed a touch-up).

**Pull request merged**
